simplic-authentication/Simplic.Authentication.Service/AuthenticationService.cs

```diff
@@ -41,45 +45,133 @@ public AuthenticationService(IUnityContainer container, IAuthorizationService au
         /// <returns>A user session if the user could be logged in, else an exception will be thrown</returns>
         public Session Login(string domain, string userName, string password)
         {
-            Session session = null;
-
             if (string.IsNullOrWhiteSpace(userName))
                 throw new LoginFailedException(LoginFailedType.UserNameNotEntered);
 
             if (string.IsNullOrWhiteSpace(password))
-                throw new LoginFailedException(LoginFailedType.PasswordNotEntered);
+                throw new LoginFailedException(LoginFailedType.PasswordNotEntered);
+
+            var user = userService.GetByName(userName);
+
+            var providerName = user.IsADUser ? "ActiveDirectoryCredentialProvider" : "DefaultCredentialProvider";
+            var provider = container.Resolve<ICredentialProvider>(providerName);
 
-            var user = userService.GetByName(userName);
+            if (!provider.CheckCredentials(domain, userName, password, user.Password))
+                throw new LoginFailedException(LoginFailedType.InvalidCredentials);
 
+            return GenerateUserSession(user);
+        }
+
+        /// <summary>
+        /// Generate user session by current user instance
+        /// </summary>
+        /// <param name="user">User instance</param>
+        /// <returns>UserSession object if valid</returns>
+        private Session GenerateUserSession(User.User user)
+        {
             if (user == null)
                 throw new LoginFailedException(LoginFailedType.UserNotFound);
 
             if (!user.IsActive)
                 throw new LoginFailedException(LoginFailedType.UserNotActive);
 
-            var providerName = user.IsADUser ? "ActiveDirectoryCredentialProvider" : "DefaultCredentialProvider";
-            var provider = container.Resolve<ICredentialProvider>(providerName);
-
-            if (!provider.CheckCredentials(domain, userName, password, user.Password))
-                throw new LoginFailedException(LoginFailedType.InvalidCredentials);
-
             var userGroups = groupService.GetAllByUserId(user.Ident);
             var isSuperUser = userGroups.Any(x => x.GroupId == 0);
             var userGroupIdents = userGroups.Select(x => x.Ident).ToList();
 
-            session = new Session()
+            return new Session()
```

Comment:
you dont need parenthesis if you assign variables directly after creating them. This code could be simplified as follows:
```csharp
return new Session {
    UserId = user.Ident,
    ...
};
```